elastic/apm-server
# Add initial support for indexing to data streams

## Motivation/summary

This PR introduces initial support for indexing events into data streams, with the new `apm-server.data_streams.enabled` config. When this is set, the server will:
 - add `data_stream.{type,dataset,namespace}` fields to each published event
 - use these fields to route the events to the appropriate data stream (`$type-$dataset-$namespace`)
 - disable ILM setup by default

Still TODO, in followups:
 - system tests
 - set `apm-server.data_streams.enabled` when in Fleet mode
 - make `data_stream.namespace` configurable (using config received from Elastic Agent)
 - improve config validation, to ensure things like `output.elasticsearch.indices` cannot be set in conjunction `apm-server.data_streams.enabled`

## Checklist


I have considered changes for:
~- [ ] documentation~ (later maybe; more likely this will be undocumented and set indirectly by using Fleet)
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

- Run apm-server with `-E apm-server.data_streams.enabled=true`
- Send some transactions, spans, errors, and runtime metrics
- Confirm that transactions and spans go into `traces-*`, errors into `logs-*`, and runtime metrics into `metrics-*`.

## Related issues

Partially addresses #4378
